cmd/zoekt-sourcegraph-indexserver/sg.go

```diff
@@ -128,24 +168,53 @@ type sourcegraphClient struct {
 	// zero a value of 10000 is used.
 	BatchSize int
 
-	// Client is used to make requests to the Sourcegraph instance. Prefer to
+	// restClient is used to make requests to the Sourcegraph instance. Prefer to
 	// use .doRequest() to ensure the appropriate headers are set.
-	Client *retryablehttp.Client
+	restClient *retryablehttp.Client
+
+	// grpcClient is used to make requests to the Sourcegraph instance if gRPC is enabled.
+	grpcClient proto.IndexedSearchConfigurationServiceClient
 
 	// configFingerprint is the last config fingerprint returned from
 	// Sourcegraph. It can be used for future calls to the configuration
 	// endpoint.
 	configFingerprint string
 
+	// configFingerprintProto is the last config fingerprint (as GRPC) returned from
+	// Sourcegraph. It can be used for future calls to the configuration
+	// endpoint.
+	configFingerprintProto *proto.Fingerprint
```

Comment:
I thought about this for a while, but the simplest solution that I can come up with to implement the fingerprinting logic for both gRPC and protobuf was  to duplicate a lot of the logic:

- Create an abstraction (`getIndexOptionsFunc`) to hide the exact type of the fingerprint that's being managed
- Create a separate field for each type of fingerprint (string for REST, protobuf for gRPC)

I don't think is great, but I thought that introducing a separate interface wasn't worth it since this only has one user: the `List()` function in the Sourcegraph Client. I'd appreciate other ideas though.

Reply:
The comment here is a little confusing. Am I understanding that `configFingerprintProto` and `configFingerprint` are mutually exclusive? And `configFingerprintProto` is only used if gRPC is enabled?

Reply:
Discussed offline. These fingerprints are two different types. Doc comments have been added to clarify this.